src/txn/txn_rollback_to_stable.c

```diff
@@ -201,66 +201,78 @@ __txn_abort_newer_row_leaf(
     }
 }
 
+/*
+ * __txn_page_needs_rollback --
+ *     Check whether the page needs rollback. Return true if the page has modifications newer than
+ *     the given timestamp Otherwise return false.
+ */
+static bool
+__txn_page_needs_rollback(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t rollback_timestamp)
+{
+    WT_ADDR *addr;
+    WT_CELL_UNPACK vpack;
+    WT_MULTI *multi;
+    WT_PAGE_MODIFY *mod;
+    wt_timestamp_t multi_newest_durable_ts;
+    uint32_t i;
+
+    addr = ref->addr;
+    mod = ref->page == NULL ? NULL : ref->page->modify;
+
+    /*
+     * The rollback operation should be performed on this page when any one of the following is
+     * greater than the given timestamp:
+     * 1. The reconciled replace page max durable timestamp.
+     * 2. The reconciled multi page max durable timestamp.
+     * 3. The on page address max durable timestamp.
+     * 4. The off page address max durable timestamp.
+     */
+    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
+        return (mod->mod_replace.newest_durable_ts > rollback_timestamp);
+    else if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK) {
+        multi_newest_durable_ts = WT_TS_NONE;
+        /* Calculate the max durable timestamp by traversing all multi addresses. */
+        for (multi = mod->mod_multi, i = 0; i < mod->mod_multi_entries; ++multi, ++i)
+            multi_newest_durable_ts =
+              WT_MAX(multi_newest_durable_ts, multi->addr.newest_durable_ts);
+        return (multi_newest_durable_ts > rollback_timestamp);
+    } else if (!__wt_off_page(ref->home, addr)) {
+        /* Check if the page is obsolete using the page disk address. */
+        __wt_cell_unpack(session, ref->home, (WT_CELL *)addr, &vpack);
+        return (vpack.newest_durable_ts > rollback_timestamp);
+    } else if (addr != NULL)
+        return (addr->newest_durable_ts > rollback_timestamp);
+
+    return (false);
```

Comment:
I assume we have exclusive access to the tree at this point?

Reply:
Yes. But there are some scenarios where there is no address and the reconciliation leads to page empty. To handle them, I added those checks. Those pages don't need any rollback.